bentoml/frameworks/paddle.py

```diff
@@ -0,0 +1,103 @@
+import os
+
+from bentoml.exceptions import MissingDependencyException
+from bentoml.service.artifacts import BentoServiceArtifact
+from bentoml.service.env import BentoServiceEnv
+import tempfile
+
+
+try:
+    import paddle
+    import paddle.inference as paddle_infer
+except ImportError:
+    paddle = None
+
+
+class PaddleModelArtifact(BentoServiceArtifact):
+    """Abstraction for saving/loading PaddlePaddle models
+
+    Args:
+        name (string): name of the artifact
+
+    Raises:
+        MissingDependencyException: paddle package is required for PaddleModelArtifact
+
+    Example usage:
+
+    >>> import pandas as pd
+    >>>
+    >>> from bentoml import env, artifacts, api, BentoService
+    >>> from bentoml.adapters import DataframeInput
+    >>> from bentoml.frameworks.paddle import PaddleModelArtifact
+    >>>
+    >>> @env(infer_pip_packages=True)
+    >>> @artifacts([PaddleModelArtifact('model')])
+    >>> class PaddleService(BentoService):
+    >>>    @api(input=DataframeInput(), batch=True)
+    >>>    def predict(self, df: pd.DataFrame):
+    >>>        input_data = df.to_numpy().astype(np.float32)
+    >>>        predictor = self.artifacts.model
+    >>>
+    >>>        input_names = predictor.get_input_names()
+    >>>        input_handle = predictor.get_input_handle(input_names[0])
+    >>>        input_handle.reshape(input_data.shape)
+    >>>        input_handle.copy_from_cpu(input_data)
+    >>>
+    >>>        predictor.run()
+    >>>
+    >>>        output_names = predictor.get_output_names()
+    >>>        output_handle = predictor.get_output_handle(output_names[0])
+    >>>        output_data = output_handle.copy_to_cpu()
+    >>>        return output_data
+    >>>
+    >>> service = PaddleService()
+    >>>
+    >>> service.pack('model', model_to_save)
+    """
+
+    def __init__(self, name: str):
+        super(PaddleModelArtifact, self).__init__(name)
+        self._model = None
+        self._predictor = None
+        self._model_path = None
+
+        if paddle is None:
+            raise MissingDependencyException(
+                "paddlepaddle package is required to use PaddleModelArtifact"
+            )
+
+    def pack(self, model):  # pylint:disable=arguments-differ
+        self._model = model
+        return self
+
+    def load(self, path):
+        model = paddle.jit.load(self._file_path(path))
+        model = paddle.jit.to_static(model, input_spec=model._input_spec())
+        return self.pack(model)
+
+    def _file_path(self, base_path):
+        return os.path.join(base_path, self.name)
+
+    def save(self, dst):
+        self._save(dst)
+
+    def _save(self, dst):
+        self._model_path = self._file_path(dst)
+        paddle.jit.save(self._model, self._model_path)
+
+    def get(self):
+        if not self._predictor:
+            if self._model_path is None:
```

Comment:
Let's add an inline comment about why we are doing this.  It will be very useful for the future, if we need to looking at this again.